lib/emerald/parser.rb

```diff
@@ -6,17 +6,27 @@ def initialize(source)
 
     def parse
       ast = []
-      result = parse_node(@source)
+      parse_lists_and_nodes(@source, ast)
+    end
+
+    private
+
+    def parse_lists_and_nodes(source, ast)
+      result = parse_node(source)
       while result
         node, source = result
-        ast.push(node) if node
+        if node.is_a? List
```

Comment:
I see what you mean, but this method is funnelling how things are added to the ast, based on what's returned from the parse_node method (a node or list). It's quite tightly coupled to parsing both the lists and nodes (perhaps it shouldn't be?), so I'm not sure what could be extracted...